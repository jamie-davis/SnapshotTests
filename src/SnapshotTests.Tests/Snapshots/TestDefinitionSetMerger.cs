using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NUnit.Framework;
using SnapshotTests.Snapshots;
using TestConsoleLib;
using TestConsoleLib.Testing;

namespace SnapshotTests.Tests.Snapshots
{
    [TestFixture]
    public class TestDefinitionSetMerger
    {
        #region Types that define tables

        [SnapshotDefinition("LocalTable")]
        public static class A_TableDef
        {
            [Unpredictable] 
            public static int Id { get; set; }
        }

        [SnapshotDefinition("PredictableTable")]
        public static class A_PredictableTable
        {
            [Predictable] 
            public static int Id { get; set; }
        }

        [SnapshotDefinition("LocalTable")]
        public static class B_TableDef
        {
            [Unpredictable] 
            public static int Id { get; set; }
        }

        [SnapshotDefinition("PredictableTable")]
        public static class B_PredictableTable
        {
            [Predictable] 
            public static int Id { get; set; }
            [References("LocalTable", "Id")]
            public static int Ref { get; set; }
        }

        [SnapshotDefinition("AddedTable")]
        public static class B_AddedTable
        {
            [Predictable] 
            public static int Id { get; set; }
        }

        #endregion

        [Test]
        public void DefinitionsAreLoadedFromAssembly()
        {
            //Arrange
            var definition = SnapshotDefinitionLoader.Load(GetType(), t => t.Name.StartsWith("A_"));

            //Act
            var loaded = DefinitionSetMerger.Merge(definition, null).ToList();

            //Assert
            var output = new Output();
            TableDefinitionReporter.Report(loaded, output);
            output.Report.Verify();
        }

        [Test]
        public void DefinitionsAreMerged()
        {
            //Arrange
            var definition = SnapshotDefinitionLoader.Load(GetType(), t => t.Name.StartsWith("A_"));
            var definition2 = SnapshotDefinitionLoader.Load(GetType(), t => t.Name.StartsWith("B_"));
            var loaded = DefinitionSetMerger.Merge(definition, null).ToList();

            //Act
            var result = DefinitionSetMerger.Merge(definition2, loaded).ToList();

            //Assert
            var output = new Output();
            TableDefinitionReporter.Report(result, output);
            output.Report.Verify();
        }
    }
}
