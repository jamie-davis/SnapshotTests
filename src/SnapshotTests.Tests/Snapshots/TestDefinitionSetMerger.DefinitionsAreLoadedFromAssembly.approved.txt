Table Definitions                                                                              
-----------------                                                                              
                 Exclude                                                                       
                 From       Include In                                                         
Table Name       Comparison Comparison Defined By Types                                        
---------------- ---------- ---------- ------------------                                      
LocalTable       False      False      A_TableDef                                              
                                                                                               
         Is      Is                                                    Referenced                
         Primary Compare Is            Is          Is       Referenced Property   Sort      Sort 
    Name Key     Key     Unpredictable Predictable Required Table Name Name       Direction Index
    ---- ------- ------- ------------- ----------- -------- ---------- ---------- --------- -----
    Id   False   False   True          False       False                                         
                                                                                               
                 Exclude                                                                       
                 From       Include In                                                         
Table Name       Comparison Comparison Defined By Types                                        
---------------- ---------- ---------- ------------------                                      
PredictableTable False      False      A_PredictableTable                                      
                                                                                               
         Is      Is                                                    Referenced                
         Primary Compare Is            Is          Is       Referenced Property   Sort      Sort 
    Name Key     Key     Unpredictable Predictable Required Table Name Name       Direction Index
    ---- ------- ------- ------------- ----------- -------- ---------- ---------- --------- -----
    Id   False   False   False         True        False                                         
                                                                                               
                                                                                               